File: Leetcode74.js
// Time Complexity : O(1)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no


// Your code here along with comments explaining your approach: Simply just concatenated the arrays inside base array into base array and then searched with linear approach and using indexOf().


/**
 * @param {number[][]} matrix
 * @param {number} target
 * @return {boolean}
 */
var searchMatrix = function(matrix, target) {
    const arr = [].concat(...matrix);
    return arr.indexOf(target)!==-1?true:false;
};

File: Leetcode74.java
class Leetcode74 {
    public boolean searchMatrix(int[][] matrix, int target) {
        if (matrix == null || matrix.length == 0)
            return false;

        int r = matrix.length;
        int c = matrix[0].length;
        int l = 0;
        int h = r * c - 1;

        while (l <= h) {
            int mid = l + (h - l) / 2;
            int row = mid / c;
            int col = mid % c;
            if (matrix[row][col] == target) {
                return true;
            } else if (target > matrix[row][col]) {
                l = mid + 1;
            } else {
                h = mid - 1;
            }
        }
        return false;
    }
}

// time complexity: O(log r*c)
// space : O(1)
// logic to find row and column in any 2d array:
// row=index/number of cols
// col=index%number of cols